# App Store Analytics
Explore thousands of apps data on Google Play store

In [1]:
import pandas as pd
import numpy as np
import plotly.express as px

## Data Exploration

In [2]:
df_apps = pd.read_csv('data/apps.csv')
df_apps.sample(n=10, replace=True, random_state=1)

App         Category  Rating  Reviews  \
235           Trovami se ci riesci             GAME     5.0       11   
5192              Castlight Mobile          MEDICAL     3.7      529   
905   California Cop Assist CA Cop     PRODUCTIVITY     3.0        5   
7813            QuickShortcutMaker  PERSONALIZATION     4.6    41000   
2895                  HCP Anywhere         BUSINESS     4.7      114   
5056                    Type S LED        LIFESTYLE     3.5      628   
144        Cy-Fair Houston Chamber         BUSINESS     NaN        0   
4225                       Cl-app!           SPORTS     NaN       41   
7751                  PJ Masks: HQ           FAMILY     4.1    13731   
3462                   Old: CL-150           FAMILY     3.8      120   

       Size_MBs   Installs  Type  Price Content_Rating  \
235    6.100000         10  Free      0       Everyone   
5192  26.000000    100,000  Free      0       Everyone   
905    7.000000        100  Paid  $4.99       Everyone   
7813   2.000000  1,000,000  Free      0       Everyone   
2895   8.600000      5,000  Free      0       Everyone   
5056   8.350000    100,000  Free      0       Everyone   
144    5.000000          5  Free      0       Everyone   
4225   0.344727     10,000  Free      0       Everyone   
7751  55.000000  1,000,000  Free      0       Everyone   
3462  24.000000     10,000  Free      0       Everyone   

                                Genres       Last_Updated   Android_Ver  
235                             Arcade     March 11, 2017    2.3 and up  
5192                           Medical      July 30, 2018    5.0 and up  
905                       Productivity  February 11, 2014    2.2 and up  
7813                   Personalization  February 23, 2014    1.6 and up  
2895                          Business     March 30, 2018    4.3 and up  
5056                         Lifestyle    August 14, 2017    4.3 and up  
144                           Business       June 6, 2018    4.1 and up  
4225                            Sports        May 2, 2013    2.0 and up  
7751  Entertainment;Action & Adventure      June 22, 2018    4.0 and up  
3462                         Education       May 21, 2018  4.0.3 and up

In [3]:
df_apps.shape

(10841, 12)

So our DataFrame has **10841** rows and **12** columns.

Let's check if there are any `NaN` values in it.

In [4]:
df_apps.isna().any()

App               False
Category          False
Rating             True
Reviews           False
Size_MBs          False
Installs          False
Type               True
Price             False
Content_Rating    False
Genres            False
Last_Updated      False
Android_Ver        True
dtype: bool

We can see that there are `NaN` values in **Rating**, **Type** and **Android_Ver** columns.

In [5]:
df_apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size_MBs        10841 non-null  float64
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content_Rating  10841 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last_Updated    10841 non-null  object 
 11  Android_Ver     10839 non-null  object 
dtypes: float64(2), int64(1), object(9)
memory usage: 1016.5+ KB


Now we are going to remove **Last_Updated** and **Android_Ver** because they are not necessary.

In [6]:
df_apps = df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1)

Let's see how many rows in the **Rating** and the **Type** columns have `NaN` values.

In [7]:
print(df_apps['Rating'].isna().value_counts())
print(df_apps['Type'].isna().value_counts())

False    9367
True     1474
Name: Rating, dtype: int64
False    10840
True         1
Name: Type, dtype: int64


There are 1474 rows in **Rating** and 1 row in **Type** have `NaN` values. Let's drop them.

In [8]:
df_apps = df_apps.dropna(axis=0)
df_apps.isna().any()

App               False
Category          False
Rating            False
Reviews           False
Size_MBs          False
Installs          False
Type              False
Price             False
Content_Rating    False
Genres            False
dtype: bool

In [9]:
df_apps.shape

(9367, 10)

We are not done yet. We have to find any duplicated apps and delete them. The way we do this is by looking up in the **App**, **Type** and **Price** columns and look for any duplicated features.

For example: let's look for any  duplicated **Instagram** in the DataFrame.

In [10]:
df_instagram = df_apps[df_apps['App'] == 'Instagram']
df_instagram

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   
10808  Instagram   SOCIAL     4.5  66577446       5.3  1,000,000,000  Free   
10809  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   
10810  Instagram   SOCIAL     4.5  66509917       5.3  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social  
10808     0           Teen  Social  
10809     0           Teen  Social  
10810     0           Teen  Social

In [11]:
df_instagram.duplicated()

10806    False
10808    False
10809     True
10810    False
dtype: bool

We can see that there 4 entries of **Instagram** but only one at index **10809** is a duplicate. If we simply call [.drop_duplicates](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html?highlight=drop_duplicate) on our DataFrame, it will only remove the one at index **10809**. 

Therefore we have to tell our DataFrame how to identify a duplicate. We have to provide the column names that we want it to look for duplicated features in the `subset` parameter, in this case, look for duplicates in the **App**, **Type** and **Price** column.

In [12]:
df_apps[df_apps.duplicated(subset=['App', 'Type', 'Price'])]

App Category  Rating   Reviews  \
946                        420 BZ Budeze Delivery  MEDICAL     5.0         2   
1133                                  MouseMingle   DATING     2.7         3   
1196   Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL     4.4         8   
1231                                 Sway Medical  MEDICAL     5.0         3   
1247               Chat Kids - Chat Room For Kids   DATING     4.7         6   
...                                           ...      ...     ...       ...   
10836                              Subway Surfers     GAME     4.5  27723193   
10837                              Subway Surfers     GAME     4.5  27724094   
10838                              Subway Surfers     GAME     4.5  27725352   
10839                              Subway Surfers     GAME     4.5  27725352   
10840                              Subway Surfers     GAME     4.5  27711703   

       Size_MBs       Installs  Type   Price Content_Rating   Genres  
946        11.0            100  Free       0     Mature 17+  Medical  
1133        3.9            100  Free       0     Mature 17+   Dating  
1196        6.5            100  Paid  $12.99       Everyone  Medical  
1231       22.0            100  Free       0       Everyone  Medical  
1247        4.9            100  Free       0     Mature 17+   Dating  
...         ...            ...   ...     ...            ...      ...  
10836      76.0  1,000,000,000  Free       0   Everyone 10+   Arcade  
10837      76.0  1,000,000,000  Free       0   Everyone 10+   Arcade  
10838      76.0  1,000,000,000  Free       0   Everyone 10+   Arcade  
10839      76.0  1,000,000,000  Free       0   Everyone 10+   Arcade  
10840      76.0  1,000,000,000  Free       0   Everyone 10+   Arcade  

[1168 rows x 10 columns]

In [13]:
df_apps = df_apps.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps[df_apps.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL     4.5  66577313       5.3  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

Now we have successfully removed duplicated rows of Instagram as well as which of other apps.

In [14]:
df_apps.shape

(8199, 10)

Now we move on.

Let's take a look which apps have highest rating, which apps have largest size in MB, which ones have greatest number of reviews.

In [15]:
# apps with highest rating
df_apps.sort_values('Rating', ascending=False).head(10)

App      Category  Rating  \
21                                 KBA-EZ Health Guide       MEDICAL     5.0   
1230                                      Sway Medical       MEDICAL     5.0   
1227                                 AJ Men's Grooming     LIFESTYLE     5.0   
1224                                    FK Dedinje BGD        SPORTS     5.0   
1223                                   CB VIDEO VISION   PHOTOGRAPHY     5.0   
1222                        Beacon Baptist Jupiter, FL     LIFESTYLE     5.0   
1214                                    BV Mobile Apps  PRODUCTIVITY     5.0   
2680                               Florida Wildflowers        FAMILY     5.0   
1206                                      ADS-B Driver         TOOLS     5.0   
2750  Superheroes, Marvel, DC, Comics, TV, Movies News        COMICS     5.0   

      Reviews  Size_MBs Installs  Type  Price Content_Rating        Genres  
21          4      25.0        1  Free      0       Everyone       Medical  
1230        3      22.0      100  Free      0       Everyone       Medical  
1227        2      22.0      100  Free      0       Everyone     Lifestyle  
1224       36       2.6      100  Free      0       Everyone        Sports  
1223       13       2.6      100  Free      0       Everyone   Photography  
1222       14       2.6      100  Free      0       Everyone     Lifestyle  
1214        3       4.8      100  Free      0       Everyone  Productivity  
2680        5      69.0    1,000  Free      0       Everyone     Education  
1206        2       6.3      100  Paid  $1.99       Everyone         Tools  
2750       34      12.0    5,000  Free      0       Everyone        Comics

In [16]:
# apps with largest size in MB
df_apps.sort_values('Size_MBs', ascending=False).head(10)

App            Category  \
9942                     Talking Babsy Baby: Baby Games           LIFESTYLE   
10687                            Hungry Shark Evolution                GAME   
9943                              Miami crime simulator                GAME   
9944                       Gangster Town: Vice District              FAMILY   
3144                                         Vi Trainer  HEALTH_AND_FITNESS   
9945                                    Ultimate Tennis              SPORTS   
7926                                          Post Bank             FINANCE   
7927                        The Walking Dead: Our World                GAME   
7928                      Stickman Legends: Shadow Wars                GAME   
4176   Car Crash III Beam DH Real Damage Simulator 2018                GAME   

       Rating  Reviews  Size_MBs     Installs  Type  Price Content_Rating  \
9942      4.0   140995     100.0   10,000,000  Free      0       Everyone   
10687     4.5  6074334     100.0  100,000,000  Free      0           Teen   
9943      4.0   254518     100.0   10,000,000  Free      0     Mature 17+   
9944      4.3    65146     100.0   10,000,000  Free      0     Mature 17+   
3144      3.6      124     100.0        5,000  Free      0       Everyone   
9945      4.3   183004     100.0   10,000,000  Free      0       Everyone   
7926      4.5    60449     100.0    1,000,000  Free      0       Everyone   
7927      4.0    22435     100.0    1,000,000  Free      0           Teen   
7928      4.4    38419     100.0    1,000,000  Paid  $0.99   Everyone 10+   
4176      3.6      151     100.0       10,000  Free      0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness  
9945                   Sports  
7926                  Finance  
7927                   Action  
7928                   Action  
4176                   Racing

In [17]:
# apps with highest reviews
df_apps.sort_values('Reviews', ascending=False).head(10)

App       Category  \
10805                                           Facebook         SOCIAL   
10785                                 WhatsApp Messenger  COMMUNICATION   
10806                                          Instagram         SOCIAL   
10784           Messenger – Text and Video Chat for Free  COMMUNICATION   
10650                                     Clash of Clans           GAME   
10744            Clean Master- Space Cleaner & Antivirus          TOOLS   
10835                                     Subway Surfers           GAME   
10828                                            YouTube  VIDEO_PLAYERS   
10746  Security Master - Antivirus, VPN, AppLock, Boo...          TOOLS   
10584                                       Clash Royale           GAME   

       Rating   Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805     4.1  78158306      5.30  1,000,000,000  Free     0           Teen   
10785     4.4  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806     4.5  66577313      5.30  1,000,000,000  Free     0           Teen   
10784     4.0  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650     4.6  44891723     98.00    100,000,000  Free     0   Everyone 10+   
10744     4.7  42916526      3.40    500,000,000  Free     0       Everyone   
10835     4.5  27722264     76.00  1,000,000,000  Free     0   Everyone 10+   
10828     4.3  25655305      4.65  1,000,000,000  Free     0           Teen   
10746     4.7  24900999      3.40    500,000,000  Free     0       Everyone   
10584     4.6  23133508     97.00    100,000,000  Free     0   Everyone 10+   

                        Genres  
10805                   Social  
10785            Communication  
10806                   Social  
10784            Communication  
10650                 Strategy  
10744                    Tools  
10835                   Arcade  
10828  Video Players & Editors  
10746                    Tools  
10584                 Strategy

We can see that those apps with highest rating have very little review, not enough to regulate more accurate rating.

Apps with highest size in MB are mostly games. And also, there seems to be an upper bound of 100MB in term of size limit on Google Play.

Finally, apps with highest reviews are mostly social network apps or communication apps and coincidentally free.

## Data Visualization with Plotly
All Android apps have a content rating like “Everyone” or “Teen” or “Mature 17+”. Let’s take a look at the distribution of the content ratings in our dataset and see how to visualise it with [plotly](https://plotly.com/python/)

In [18]:
ratings = df_apps['Content_Rating'].value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

Create a pie chart

In [19]:
labels_dict = {}

for key in ratings.index:
    labels_dict[key] = key

print(labels_dict)

{'Everyone': 'Everyone', 'Teen': 'Teen', 'Mature 17+': 'Mature 17+', 'Everyone 10+': 'Everyone 10+', 'Adults only 18+': 'Adults only 18+', 'Unrated': 'Unrated'}


In [20]:
fig = px.pie(names=ratings.index, values=ratings, labels=labels_dict, title='Content Rating')

fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()

Look for apps that have over 1 billion installs. Careful, the `Install` column is not in the numeric format, indeed it is in string format because it contains `,` in its values.

In [21]:
df_apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   object 
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(1), object(7)
memory usage: 704.6+ KB


First we replace the `,` with an empty string.

Then we use [to_numeric()](https://pandas.pydata.org/docs/reference/api/pandas.to_numeric.html#pandas.to_numeric) method to convert the `Install` column to numeric format.

In [22]:
df_apps['Installs'] = df_apps['Installs'].str.replace(',', '')
df_apps['Installs'] = pd.to_numeric(df_apps['Installs'])
df_apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   int64  
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   object 
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(2), int64(2), object(6)
memory usage: 704.6+ KB


In [23]:
df_apps[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

We can see that there are 20 apps with more than 1 billion installs.

In [24]:
df_apps.sample(n=20, random_state=42)

App             Category  \
5113                         MyLifeOrganized: To-Do List         PRODUCTIVITY   
2693                          Trinomial Factoring Wizard                TOOLS   
10318                                            PAC-MAN                 GAME   
3876                      Bulgarian French Dictionary Fr  BOOKS_AND_REFERENCE   
6041                                         Manga Books               COMICS   
7158                          1800 Contacts - Lens Store              MEDICAL   
4527                      ER Blood Draw Doctor & Surgeon               FAMILY   
2940                                       Makeup Videos               BEAUTY   
1088                                     Phonics Puzzles               FAMILY   
1933                        Complete Anatomy for Android              MEDICAL   
5420   Trell: An app for Explorers | Best Wanderlust App            LIFESTYLE   
9636                      Zapya - File Transfer, Sharing                TOOLS   
7506              Mirror Camera (Mirror + Selfie Camera)               BEAUTY   
5774                                      Faustop Sounds               COMICS   
5706                                   Remix Second Hand             SHOPPING   
3129                                      MedStar eVisit              MEDICAL   
8605                                 Monitor Your Weight   HEALTH_AND_FITNESS   
10486                                 Speedtest by Ookla                TOOLS   
6842                                     Free Resume App             BUSINESS   
7216                                         Slendrina X                 GAME   

       Rating  Reviews   Size_MBs   Installs  Type  Price   Content_Rating  \
5113      4.3     4441   4.000000     100000  Free      0         Everyone   
2693      4.0       22   0.080078       1000  Paid  $0.99         Everyone   
10318     4.2   685981  37.000000  100000000  Free      0         Everyone   
3876      4.6      603   7.400000      10000  Free      0         Everyone   
6041      3.8     7326   8.050000     500000  Free      0  Adults only 18+   
7158      4.7    23160  26.000000    1000000  Free      0         Everyone   
4527      4.0      154  58.000000      50000  Free      0             Teen   
2940      3.8        9   3.400000       5000  Free      0         Everyone   
1088      3.8        4  54.000000        100  Paid  $2.99         Everyone   
1933      4.1       90  21.000000       1000  Free      0         Everyone   
5420      4.6     8004  34.000000     100000  Free      0             Teen   
9636      4.6   387958   9.000000   10000000  Free      0             Teen   
7506      4.1     9315   2.600000    1000000  Free      0         Everyone   
5774      4.7     1677  40.000000     100000  Free      0         Everyone   
5706      4.2      867   7.500000     100000  Free      0         Everyone   
3129      4.5       20  40.000000       5000  Free      0         Everyone   
8605      4.5   126017   4.300000    5000000  Free      0         Everyone   
10486     4.4  1028794   3.400000  100000000  Free      0         Everyone   
6842      4.6    15830   7.200000    1000000  Free      0         Everyone   
7216      4.4    12736  56.000000    1000000  Free      0             Teen   

                  Genres  
5113        Productivity  
2693               Tools  
10318             Arcade  
3876   Books & Reference  
6041              Comics  
7158             Medical  
4527         Educational  
2940              Beauty  
1088         Educational  
1933             Medical  
5420           Lifestyle  
9636               Tools  
7506              Beauty  
5774              Comics  
5706            Shopping  
3129             Medical  
8605    Health & Fitness  
10486              Tools  
6842            Business  
7216              Arcade

Now we move on to finding top 20 most expensive apps.

The same with the `Install` column, the `Price` column is in string format because it contains `$` character in its values. We have to remove it and convert the column into numeric type.

In [25]:
df_apps['Price'] = df_apps['Price'].str.replace('$', '')
df_apps['Price'] = pd.to_numeric(df_apps['Price'])
df_apps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8199 entries, 21 to 10835
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             8199 non-null   object 
 1   Category        8199 non-null   object 
 2   Rating          8199 non-null   float64
 3   Reviews         8199 non-null   int64  
 4   Size_MBs        8199 non-null   float64
 5   Installs        8199 non-null   int64  
 6   Type            8199 non-null   object 
 7   Price           8199 non-null   float64
 8   Content_Rating  8199 non-null   object 
 9   Genres          8199 non-null   object 
dtypes: float64(3), int64(2), object(5)
memory usage: 704.6+ KB


<ipython-input-25-faac03119c97>:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



Now, the top 20 most expensive apps.

In [26]:
df_apps.nlargest(20, 'Price')

App   Category  Rating  Reviews   Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE     3.6      275   7.300000   
1331          most expensive app (H)     FAMILY     4.3        6   1.500000   
1946  I am rich (Most expensive app)    FINANCE     4.1      129   2.700000   
2394                      I am Rich!    FINANCE     3.8       93  22.000000   
2461              I AM RICH PRO PLUS    FINANCE     4.0       36  41.000000   
2775                   I Am Rich Pro     FAMILY     4.4      201   2.700000   
3114                       I am Rich    FINANCE     4.3      180   3.800000   
3145              I am rich(premium)    FINANCE     3.5      472   0.942383   
3221                  I am Rich Plus     FAMILY     4.0      856   8.700000   
3554                      💎 I'm rich  LIFESTYLE     3.8      718  26.000000   
4606               I Am Rich Premium    FINANCE     4.1     1867   4.700000   
5765                       I am rich  LIFESTYLE     3.8     3547   1.800000   
3897                       I Am Rich     FAMILY     3.6      217   4.900000   
2193             I am extremely Rich  LIFESTYLE     2.9       41   2.900000   
3856                   I am rich VIP  LIFESTYLE     3.8      411   2.600000   
2281       Vargo Anesthesia Mega App    MEDICAL     4.6       92  32.000000   
1407                    LTC AS Legal    MEDICAL     4.0        6   1.300000   
2629                I am Rich Person  LIFESTYLE     4.2      134   1.800000   
2481         A Manual of Acupuncture    MEDICAL     3.5      214  68.000000   
504        AP Art History Flashcards     FAMILY     5.0        1  96.000000   

      Installs  Type   Price Content_Rating         Genres  
3946     10000  Paid  400.00       Everyone      Lifestyle  
1331       100  Paid  399.99       Everyone  Entertainment  
1946      1000  Paid  399.99           Teen        Finance  
2394      1000  Paid  399.99       Everyone        Finance  
2461      1000  Paid  399.99       Everyone        Finance  
2775      5000  Paid  399.99       Everyone  Entertainment  
3114      5000  Paid  399.99       Everyone        Finance  
3145      5000  Paid  399.99       Everyone        Finance  
3221     10000  Paid  399.99       Everyone  Entertainment  
3554     10000  Paid  399.99       Everyone      Lifestyle  
4606     50000  Paid  399.99       Everyone        Finance  
5765    100000  Paid  399.99       Everyone      Lifestyle  
3897     10000  Paid  389.99       Everyone  Entertainment  
2193      1000  Paid  379.99       Everyone      Lifestyle  
3856     10000  Paid  299.99       Everyone      Lifestyle  
2281      1000  Paid   79.99       Everyone        Medical  
1407       100  Paid   39.99       Everyone        Medical  
2629      1000  Paid   37.99       Everyone      Lifestyle  
2481      1000  Paid   33.99       Everyone        Medical  
504         10  Paid   29.99     Mature 17+      Education

Unbelievable, there are some apps that are ridiculously expensive that top **$400**.

The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for $999.90. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. 

Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well.

Leaving this bad data in our dataset will misrepresent our analysis of the most expensive 'real' apps. We might as well remove these apps.

In [27]:
df_apps = df_apps[df_apps['Price'] < 200]
df_apps.nlargest(20, 'Price')

App     Category  Rating  \
2281                          Vargo Anesthesia Mega App      MEDICAL     4.6   
1407                                       LTC AS Legal      MEDICAL     4.0   
2629                                   I am Rich Person    LIFESTYLE     4.2   
2481                            A Manual of Acupuncture      MEDICAL     3.5   
504                           AP Art History Flashcards       FAMILY     5.0   
2207                                           EMT PASS      MEDICAL     3.4   
2463                                 PTA Content Master      MEDICAL     4.2   
4264                            Golfshot Plus: Golf GPS       SPORTS     4.1   
3241  Muscle Premium - Human Anatomy, Kinesiology, B...      MEDICAL     4.2   
4772   Human Anatomy Atlas 2018: Complete 3D Human Body      MEDICAL     4.5   
526                          USMLE Step 2 CK Flashcards       FAMILY     5.0   
2119                                         NewTek NDI  PHOTOGRAPHY     3.5   
2293                               Hospitalist Handbook      MEDICAL     4.8   
4470                                  DRAGON QUEST VIII       FAMILY     4.5   
2473                            boattheory.ch Full 2018       FAMILY     4.7   
4090                             I am Rich Premium Plus      FINANCE     4.6   
1508                                 SkyTest BU/GU Lite     BUSINESS     2.9   
3778                            The World Ends With You         GAME     4.6   
2603                         2017 EMRA Antibiotic Guide      MEDICAL     4.4   
3439                            Trine 2: Complete Story         GAME     3.8   

      Reviews  Size_MBs  Installs  Type  Price Content_Rating        Genres  
2281       92      32.0      1000  Paid  79.99       Everyone       Medical  
1407        6       1.3       100  Paid  39.99       Everyone       Medical  
2629      134       1.8      1000  Paid  37.99       Everyone     Lifestyle  
2481      214      68.0      1000  Paid  33.99       Everyone       Medical  
504         1      96.0        10  Paid  29.99     Mature 17+     Education  
2207       51       2.4      1000  Paid  29.99       Everyone       Medical  
2463       64      41.0      1000  Paid  29.99       Everyone       Medical  
4264     3387      25.0     50000  Paid  29.99       Everyone        Sports  
3241      168      25.0     10000  Paid  24.99       Everyone       Medical  
4772     2921      25.0    100000  Paid  24.99       Everyone       Medical  
526         1      40.0        10  Paid  19.99       Everyone     Education  
2119       77       1.2      1000  Paid  19.99       Everyone   Photography  
2293       12      18.0      1000  Paid  19.99       Everyone       Medical  
4470     7812      27.0     50000  Paid  19.99   Everyone 10+  Role Playing  
2473       54      50.0      1000  Paid  19.40       Everyone     Education  
4090      459       2.0     10000  Paid  18.99       Everyone       Finance  
1508       28      20.0       500  Paid  17.99       Everyone      Business  
3778     4108      13.0     10000  Paid  17.99   Everyone 10+        Arcade  
2603       12       3.8      1000  Paid  16.99       Everyone       Medical  
3439      252      11.0     10000  Paid  16.99           Teen        Action

If you were to release an app, would you choose to go after a competitive category with many other apps? Or would you target a popular category with a high number of downloads? Or perhaps you can target a category which is both popular but also one where the downloads are spread out among many different apps. That way, even if it’s more difficult to be discovered among all the other apps, your app has a better chance of getting installed, right? Let’s analyse this with bar charts and scatter plots and figure out which categories are dominating the market.

In [28]:
df_apps['Category'].nunique()

33

There are 33 different categories.

Now we need to find out which category has the most apps.

In [29]:
df_most_apps_cat = df_apps['Category'].value_counts().head(10)
df_most_apps_cat

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

In [30]:
bar = px.bar(x=df_most_apps_cat.index,
             y=df_most_apps_cat.values)

bar.update_layout(xaxis_title='Category', yaxis_title='', title={'x': 0.5,
                                                                 'y': 0.95,
                                                                 'xanchor': 'center',
                                                                 'yanchor': 'top',
                                                                 'text': 'Categories with most Apps'})
bar.show()


Now we find the categories with the most installs.

In [31]:
df_most_install_cat = df_apps[['Category', 'Installs']].groupby('Category').sum()
df_most_install_cat.sort_values('Installs', ascending=True, inplace=True)
df_most_install_cat

Installs
Category                        
EVENTS                  15949410
BEAUTY                  26916200
PARENTING               31116110
MEDICAL                 39162676
COMICS                  44931100
LIBRARIES_AND_DEMO      52083000
AUTO_AND_VEHICLES       53129800
HOUSE_AND_HOME          97082000
ART_AND_DESIGN         114233100
DATING                 140912410
FOOD_AND_DRINK         211677750
EDUCATION              352852000
WEATHER                361096500
FINANCE                455249400
MAPS_AND_NAVIGATION    503267560
LIFESTYLE              503611120
BUSINESS               692018120
SPORTS                1096431465
HEALTH_AND_FITNESS    1134006220
SHOPPING              1400331540
PERSONALIZATION       1532352930
BOOKS_AND_REFERENCE   1665791655
ENTERTAINMENT         2113660000
NEWS_AND_MAGAZINES    2369110650
TRAVEL_AND_LOCAL      2894859300
VIDEO_PLAYERS         3916897200
FAMILY                4437554490
PHOTOGRAPHY           4649143130
SOCIAL                5487841475
PRODUCTIVITY          5788070180
TOOLS                 8099724500
COMMUNICATION        11039241530
GAME                 13858762717

In [32]:
bar_h = px.bar(x=df_most_install_cat.Installs,
               y=df_most_install_cat.index,
               title='Category Popularity',
               orientation='h')

bar_h.update_layout(xaxis_title='Installs',
                    yaxis_title='Categories')
bar_h.show()


From the plot above, we can see that `GAME` and `COMMUNICATION` have the most installs. 

If we plot the number of installs of a category next to the number of apps in that category we can get an idea of how concentrated a category is. Do few apps have most of the downloads or are the downloads spread out over many apps?

In [33]:
df_app_installs = df_apps[['Category', 'App', 'Installs']].groupby('Category').agg({'App': pd.Series.count,
                                                                                    'Installs': pd.Series.sum, })
df_app_installs


App     Installs
Category                              
ART_AND_DESIGN         61    114233100
AUTO_AND_VEHICLES      73     53129800
BEAUTY                 42     26916200
BOOKS_AND_REFERENCE   169   1665791655
BUSINESS              262    692018120
COMICS                 54     44931100
COMMUNICATION         257  11039241530
DATING                134    140912410
EDUCATION             118    352852000
ENTERTAINMENT         102   2113660000
EVENTS                 45     15949410
FAMILY               1606   4437554490
FINANCE               296    455249400
FOOD_AND_DRINK         94    211677750
GAME                  910  13858762717
HEALTH_AND_FITNESS    243   1134006220
HOUSE_AND_HOME         62     97082000
LIBRARIES_AND_DEMO     64     52083000
LIFESTYLE             297    503611120
MAPS_AND_NAVIGATION   118    503267560
MEDICAL               292     39162676
NEWS_AND_MAGAZINES    204   2369110650
PARENTING              50     31116110
PERSONALIZATION       298   1532352930
PHOTOGRAPHY           263   4649143130
PRODUCTIVITY          301   5788070180
SHOPPING              180   1400331540
SOCIAL                203   5487841475
SPORTS                260   1096431465
TOOLS                 719   8099724500
TRAVEL_AND_LOCAL      187   2894859300
VIDEO_PLAYERS         148   3916897200
WEATHER                72    361096500

Note: Each **dot** on the chart represents each **category**. The `size` parameter is set to `App` means that the more apps each category has the bigger the dot for that category. The `color` parameter that is set to `Installs` will toggle the color bar on the right of the chart, this means that the number of installs of each category correspond to the color on the color bar.

In [34]:
scatter = px.scatter(data_frame=df_app_installs,
                     x='App',
                     y='Installs',
                     title='Category Concentration',
                     log_y=True,
                     size='App',
                     hover_name=df_app_installs.index,
                     color='Installs')

scatter.update_layout(xaxis_title='Number of apps (Lower = more concentrated)')
scatter.show()


### Working with columns with nested data
Let’s turn our attention to the Genres column. This is quite similar to the categories column but more granular.

In [35]:
df_apps.Genres.nunique()

114

There are 114 different genres, but this is not quite accurate if we have nested data like we do in the `Genres` column.

We can see this in the below cell using `.value_counts()` and looking at the values that just have a single entry. There we see that the **semi-colon (;)** separates the genre names.

In [36]:
df_apps.Genres.value_counts().sort_values(ascending=True)[:11]

Lifestyle;Pretend Play                 1
Strategy;Education                     1
Adventure;Education                    1
Role Playing;Brain Games               1
Tools;Education                        1
Casual;Music & Video                   1
Board;Pretend Play                     1
Health & Fitness;Action & Adventure    1
Arcade;Pretend Play                    1
Strategy;Creativity                    1
Parenting;Brain Games                  1
Name: Genres, dtype: int64

We somehow need to separate the genre names to get a clear picture. This is where the string’s `.split()` method comes in handy. After we’ve separated our genre names based on the semi-colon, we can add them all into a single column with `.stack()` and then use `.value_counts()`.

In [37]:
stack = df_apps['Genres'].str.split(';', expand=True).stack()

print(stack)
print(f'Now we have a new column with shape: {stack.shape}')
print(f'Number of genres: {stack.nunique()}')


21     0                    Medical
28     0                     Arcade
47     0                     Arcade
82     0                     Arcade
99     0                    Medical
                     ...           
10824  0               Productivity
10828  0    Video Players & Editors
10829  0    Video Players & Editors
10831  0           News & Magazines
10835  0                     Arcade
Length: 8564, dtype: object
Now we have a new column with shape: (8564,)
Number of genres: 53


In [38]:
num_genres = stack.value_counts()
num_genres

Tools                      719
Education                  587
Entertainment              498
Action                     304
Productivity               301
Personalization            298
Lifestyle                  298
Finance                    296
Medical                    292
Sports                     270
Photography                263
Business                   262
Communication              258
Health & Fitness           245
Casual                     216
News & Magazines           204
Social                     203
Simulation                 200
Travel & Local             187
Arcade                     185
Shopping                   180
Books & Reference          171
Video Players & Editors    150
Dating                     134
Puzzle                     124
Maps & Navigation          118
Role Playing               111
Racing                     103
Action & Adventure          96
Strategy                    95
Food & Drink                94
Educational                 93
Adventur

In [39]:
bar_genres = px.bar(x=num_genres.index[:15],
                    y=num_genres.values[:15],
                    color=num_genres.values[:15],
                    hover_name=num_genres.index[:15],
                    color_continuous_scale='agsunset',
                    title='Top Genres')
bar_genres.update_layout(xaxis_title='Genre', yaxis_title='Number of Apps')
bar_genres.update_traces(hoverinfo='y')
bar_genres.show()
